<?xml version="1.0"?>
<robot name="lfrobot" xmlns:xacro="http://www.ros.org/wiki/xacro">
	<!-- Base link -->
	<link name="base_link">
		<visual>
			<geometry>
				<box size="0.2 0.15 0.05"/>
			</geometry>
			<material name="blue_base">
				<color rgba="0 0 1 1"/>
			</material>
		</visual>
		<collision>
			<geometry>
				<box size="0.2 0.15 0.05"/>
			</geometry>
		</collision>
		<inertial>
			<mass value="1.0"/>
			<inertia ixx="0.01" ixy="0.0" ixz="0.0"
				iyy="0.01" iyz="0.0"
				izz="0.01"/>
		</inertial>
	</link>

	<!-- xacro macro: Left and right wheels -->
	<xacro:macro name="front_wheel" params="link_name">
		<link name="${link_name}">
			<visual>
				<geometry>
					<cylinder radius="0.05" length="0.018"/>
				</geometry>
				<origin rpy="1.5708 0 0"/>
				<material name="outer_wheel">
					<color rgba="0 0 0 1"/>
				</material>
			</visual>
			<visual>
				<geometry>
					<cylinder radius="0.03" length="0.02"/>
				</geometry>
				<origin rpy="1.5708 0 0"/>
				<material name="inner_wheel">
					<color rgba="1 1 0 1"/>
				</material>
			</visual>
			<collision>
				<geometry>
					<cylinder radius="0.05" length="0.02"/>
				</geometry>
				<origin rpy="1.5708 0 0"/>
			</collision>
			<inertial>
				<mass value="0.1"/>
				<inertia ixx="0.0001" ixy="0.0" ixz="0.0"
					iyy="0.0001" iyz="0.0"
					izz="0.0001"/>
			</inertial>
		</link>
	</xacro:macro>

	<!-- Left wheel -->
	<xacro:front_wheel link_name="left_wheel"/>
	<xacro:front_wheel link_name="right_wheel"/>

	<!-- xacro macro: Left and right cameras -->
	<xacro:macro name="camera_link" params="link_name">
		<link name="${link_name}">
			<visual>
				<geometry>
					<box size="0.02 0.02 0.02"/>
				</geometry>
				<material name="camera_material">
					<color rgba="0.2 0.2 0.2 1.0"/>
				</material>
			</visual>
			<collision>
				<geometry>
					<box size="0.02 0.02 0.02"/>
				</geometry>
			</collision>
			<inertial>
				<mass value="0.05"/>
				<inertia ixx="0.00001" ixy="0.0" ixz="0.0"
					iyy="0.00001" iyz="0.0"
					izz="0.00001"/>
			</inertial>
		</link>
	</xacro:macro>

	<!-- Left and right cameras -->
	<xacro:camera_link link_name="left_camera_link"/>
	<xacro:camera_link link_name="middle_camera_link"/>
	<xacro:camera_link link_name="right_camera_link"/>

	<!-- Caster wheel -->
	<link name="caster_wheel">
		<visual>
			<geometry>
				<cylinder radius="0.0125" length="0.02"/>
			</geometry>
			<origin rpy="1.5708 0 0"/>
			<material name="caster_wheel_material">
				<color rgba="0.8 0.8 0.8 1.0"/>
			</material>
		</visual>
		<collision>
			<geometry>
				<cylinder radius="0.0125" length="0.02"/>
			</geometry>
			<origin rpy="1.5708 0 0"/>
		</collision>
		<inertial>
			<mass value="0.1"/>
			<inertia ixx="0.0001" ixy="0.0" ixz="0.0"
				iyy="0.0001" iyz="0.0"
				izz="0.0001"/>
		</inertial>
	</link>

	<!-- Left wheel joint -->
	<joint name="left_wheel_joint" type="continuous">
		<parent link="base_link"/>
		<child link="left_wheel"/>
		<origin xyz="0.03 0.085 0.0"/>
		<axis xyz="0 1 0"/>
	</joint>

	<!-- Right wheel joint -->
	<joint name="right_wheel_joint" type="continuous">
		<parent link="base_link"/>
		<child link="right_wheel"/>
		<origin xyz="0.03 -0.085 0.0"/>
		<axis xyz="0 1 0"/>
	</joint>

	<!-- Caster wheel joint -->
	<joint name="caster_wheel_joint" type="continuous">
		<parent link="base_link"/>
		<child link="caster_wheel"/>
		<origin xyz="-0.085 0 -0.0375"/>
		<axis xyz="0 1 0"/>
	</joint>

	<!-- xacro macro: Camera joint -->
	<xacro:macro name="camera_joint" params="joint_name child_link_name y_coord">
		<joint name="${joint_name}" type="fixed">
			<parent link="base_link"/>
			<child link="${child_link_name}"/>
			<origin xyz="0.1 ${y_coord} -0.03" rpy="0 1.57 0"/>
		</joint>
	</xacro:macro>

	<!-- Camera joints -->
	<xacro:camera_joint joint_name="left_camera_joint" child_link_name="left_camera_link" y_coord="0.07"/>
	<xacro:camera_joint joint_name="middle_camera_joint" child_link_name="middle_camera_link" y_coord="0.0"/>
	<xacro:camera_joint joint_name="right_camera_joint" child_link_name="right_camera_link" y_coord="-0.07"/>

	<!-- xacro macro: Camera sensor -->
	<xacro:macro name="camera_sensor" params="link_name sensor_name topic_name">
		<gazebo reference="${link_name}">
			<pose> 0 0 0 0 0 0 </pose>
			<sensor name="${sensor_name}" type="camera">
				<always_on>true</always_on>
				<topic>${topic_name}</topic>
				<visualize>true</visualize>
				<update_rate>30</update_rate>
				<camera>
					<horizontal_fov>2.0</horizontal_fov>
					<image>
			        	<width>320</width>
			        	<height>240</height>
			        	<format>R8G8B8</format>
			        </image>
		        	<clip>
		        		<near>0.001</near>
						<far>1.0</far>
		    		</clip>
		    	</camera>
			</sensor>
		</gazebo>
	</xacro:macro>

	<!-- Camera sensors -->
	<xacro:camera_sensor
		link_name="left_camera_link"
		sensor_name="left_camera"
		topic_name="left_camera/image_raw"/>
	<xacro:camera_sensor
		link_name="middle_camera_link"
		sensor_name="middle_camera"
		topic_name="middle_camera/image_raw"/>
	<xacro:camera_sensor
		link_name="right_camera_link"
		sensor_name="right_camera"
		topic_name="right_camera/image_raw"/>

	<gazebo>
		<!-- DiffDrive plugin -->
		<plugin
			filename="gz-sim-diff-drive-system"
			name="gz::sim::systems::DiffDrive">
			<left_joint>left_wheel_joint</left_joint>
			<right_joint>right_wheel_joint</right_joint>
			<wheel_separation>0.15</wheel_separation>
			<wheel_radius>0.05</wheel_radius>
			<odom_publish_frequency>1</odom_publish_frequency>
			<topic>cmd_vel</topic>
		</plugin>
	</gazebo>
</robot>
